### PR TITLE
feat: add helia/verified-fetch browser example

### PR DESCRIPTION
**Pull request by 2color**
This PR adds a new example of the @helia/verified-fetch library in the browser.

### CODE REVIEW

**Comment by SgtPooki** [comment]
@2color I added some updates in https://github.com/ipfs-examples/helia-examples/pull/286 that make using it a little more intuitive.

I'm good with these changes if we can agree on some of the changes in 286, but we should make sure linting works and tests are implemented before merging.

P.S. monorepo tests have been failing for a while

File: examples/helia-browser-react-verified-fetch/package.json [hunk 1]
```diff
@@ -0,0 +1,33 @@
+{
+  "name": "helia-browser-react-verified-fetch",
```
Comment (achingbrain) [inline examples/helia-browser-react-verified-fetch/package.json:2]: Maybe a simpler name?

```suggestion
  "name": "helia-verified-fetch",
```

**Comment by achingbrain** [comment]
This looks pretty neat, some things to fix before merging, in decreasing importance:

1. Tests - please check out how the [helia-vite example is tested](https://github.com/ipfs-examples/helia-examples/blob/main/examples/helia-vite/test/index.spec.js) - this should give you enough to crib from that the example loads and clicking the buttons works
2. The monorepo build is failing - I think this is because there are now multiple versions of eslint plugins present and they are interfering with each other - you may need to downgrade or remove something
3. Libraries - there's a lot in here that doesn't show how to use `@helia/verified-fetch` - vite, tailwind, react, postcss, etc.  Put yourself in the mind of the time-poor developer, if they don't need it to use `@helia/verified-fetch`, can it be removed?
4. Visuals - it would be most excellent to have a coherent visual identity for these examples.  Most of them use the assets from https://github.com/ipfs-shipyard/helia-css - it would be great to continue that in new additions here

File: examples/helia-browser-react-verified-fetch/package.json [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  "version": "0.0.0",
+  "type": "module",
+  "scripts": {
+    "dev": "vite",
+    "build": "tsc && vite build",
+    "lint": "eslint . --ext ts,tsx --report-unused-disable-directives --max-warnings 0",
+    "preview": "vite preview"
+  },
+  "dependencies": {
+    "@helia/verified-fetch": "0.0.0-3851fe2",
```
Comment (achingbrain) [inline examples/helia-browser-react-verified-fetch/package.json:13]: To get the latest RC until every time until we publish v1

```suggestion
    "@helia/verified-fetch": "next",
```

File: examples/helia-browser-react-verified-fetch/package.json (hunk 1 above)
Comment (2color) [reply examples/helia-browser-react-verified-fetch/package.json:2]: I'd like to keep `browser` in the name. That's a very important distinction for users.

**Comment by 2color** [comment]
@achingbrain Agree on the first two points.

Re 3: 
> Libraries - there's a lot in here that doesn't show how to use @helia/verified-fetch - vite, tailwind, react, postcss, etc. Put yourself in the mind of the time-poor junior developer, if they don't need it to use @helia/verified-fetch, can it be removed to reduce their cognitive load (and our maintenance burden)?

Yeah that's a good point. It's really hard to just use React & Tailwind without importing a whole morass of other stuff. Practically though, none of our IPFS stuff is friendly for time-poor junior developer.  

I'll take a look at:
- Stripping this to the bare minimum necessary 
- Maybe swapping Vite for esbuild for a simpler setup 

I still think it'd be useful to have an example like this (in its current state). Maybe we move some of this to the next.js example which we already have and has React. 

> Visuals - it would be most excellent to have a coherent visual identity for these examples. Some of them [use the assets](https://github.com/ipfs-examples/helia-examples/blob/main/examples/helia-esbuild/src/index.html#L9-L190) from https://github.com/ipfs-shipyard/helia-css - it would be great to continue that in new additions here

How critical is that? These are examples, why not make them a little fun and colourful?

**Comment by achingbrain** [comment]
> I still think it'd be useful to have an example like this (in its current state).

My concern with this is that we merge it and move on because we are time pressed, then six months down the line there's a [regression](https://github.com/ipfs-examples/helia-examples/issues/289) in a dep that on balance we could show how `@helia/verified-fetch` works in a browser without, then we (I) burn even more time investigating and fixing it.

This has happened multiple times in this repo and the preceding `js-ipfs-examples` one, due to how many competing UX frameworks are here, how precarious their dependency trees are and how lax they are about release breaking changes.

> Maybe we move some of this to the next.js example which we already have and has React

The focus of the `helia-nextjs` example is on any custom config you need for next.js, rather than showcasing the Helia API specifically.

The coding conventions/module formats etc between `helia` and `@helia/verified-fetch` (ESM, exports maps, etc) are the same so I'm not sure we'd be adding anything to `helia-nextjs` by doing this.

This example is about `@helia/verified-fetch` specifically so we should be showcasing the API here.

> How critical is that? These are examples, why not make them a little fun and colourful?

I'm down for fun and colourful, stick a hot pink Helia logo in there!  But again the maintenance burden is what concerns me.  If we end up with 15-20 examples in this repo and are all a bit different, trying to update things when they break can become overwhelming and people tend to give up.

File: examples/helia-browser-react-verified-fetch/package.json (hunk 1 above)
Comment (achingbrain) [reply examples/helia-browser-react-verified-fetch/package.json:2]: Ok, leave `browser` in then.

**Comment by 2color** [comment]
@achingbrain I've pushed some changes to simplify the code a fair bit and to remove as many deps as possible. 

I saw that the linting failed but I couldn't figure out exactly why: https://github.com/ipfs-examples/helia-examples/actions/runs/7843490975/job/21403945128#step:6:29

I'm getting a different error when running the `npm run lint` at the root of the repo on my local machine. Any pointers as to why?

**Comment by 2color** [comment]
@achingbrain  I added a test that checks the basic rendering of the page. It seems that all the other tests avoid any network activity for the test (sensible). 

Is there anything you'd test here other than some UI elements getting rendered?

**Comment by achingbrain** [comment]
I added a few more tests to #290 - it seems DAG-CBOR handling is a little wonky, calling `.json` on the response results in `"[object Object] is not valid JSON"`, see https://github.com/ipfs/helia/pull/426 for more.

**Comment by 2color** [comment]
@achingbrain https://github.com/ipfs/helia/pull/429 broke the tests here (https://github.com/ipfs-examples/helia-examples/actions/runs/7871416533/job/21474667989):

>Error: [vite:load-fallback] Could not load /home/runner/work/helia-examples/helia-examples/examples/helia-browser-verified-fetch/test/blockstore.js/identity (imported by ../../node_modules/@helia/utils/dist/src/utils/networked-storage.js): ENOTDIR: not a directory, open '/home/runner/work/helia-examples/helia-examples/examples/helia-browser-verified-fetch/test/blockstore.js/identity'
 
I'm looking into the helia source to see how to adapt the blockstore override for the tests to pass again. Let me know if you have any pointers.

**Comment by achingbrain** [comment]
Ha, it broke the hack that allowed testing this example.

File: examples/helia-browser-verified-fetch/test/vite.config.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
+import { resolve } from 'path'
+import defaultConfig from '../vite.config.js'
+
+// override resolution of `blockstore-core` module so we can pre-fill a memory
+// blockstore with test data
+defaultConfig.resolve ??= {}
+defaultConfig.resolve.alias ??= {}
+defaultConfig.resolve.alias['blockstore-core'] = resolve(process.cwd(), 'test/blockstore.js')
```
Comment (achingbrain) [inline examples/helia-browser-verified-fetch/test/vite.config.js:8]:
```suggestion
defaultConfig.resolve.alias['blockstore-core/dist/src/memory.js'] = resolve(process.cwd(), 'test/blockstore.js')
```

**Comment by 2color** [comment]
@achingbrain Friendly ping

**Comment by achingbrain** [comment]
I tried to load the Uniswap JSON file, nothing much was happening so I tried to load the video, then the Uniswap file appeared.

I guess in-flight requests are not cancelled when a new request is made?

File: examples/helia-browser-verified-fetch/test/index.spec.js [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+    const output = await page.locator(fetchOutput)
+    await expect(output).toContainText(
+      '{ "hello": "world" }',
+      { timeout: 2000 }
+    )
+  })
+
+  // TODO: DAG-CBOR is broken - https://github.com/ipfs/helia/pull/426
+  test.skip('should properly render ui with the ipfs path input and display DAG-CBOR', async ({ page }) => {
```
Comment (achingbrain) [inline examples/helia-browser-verified-fetch/test/index.spec.js:50]: This should be runnable now?

```suggestion
  test('should properly render ui with the ipfs path input and display DAG-CBOR', async ({ page }) => {
```

**Review by achingbrain — APPROVED** [review]
LGTM, just needs the skipped DAG-CBOR test enabling and for multiple load requests to not trample all over each other (see [comment](https://github.com/ipfs-examples/helia-examples/pull/285#issuecomment-1999966360) above)

**Comment by 2color** [comment]
I tried to add request cancellation using an AbortController but it seems to not work due to a [bug in @helia/verified-fetch](https://github.com/ipfs/helia-verified-fetch/issues/25).

**Comment by 2color** [comment]
> I guess in-flight requests are not cancelled when a new request is made?

I implemented that now.

File: examples/helia-browser-verified-fetch/src/App.tsx [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+              onClick={onAbort}
+            >
+              ❌ Abort request
+            </button>
+
+            <pre className="bg-black text-teal-300 rounded p-4 whitespace-pre-wrap break-words">{helpText}</pre>
+            <a href="https://github.com/ipfs-examples/helia-examples/tree/main/examples/helia-browser-verified-fetch" className="text-xl block mt-2 underline">Source for example</a>
+
+          </div>
+          {/* Left */}
```
Comment (SgtPooki) [inline examples/helia-browser-verified-fetch/src/App.tsx:265]:
```suggestion
```
? or `end Left`

File: examples/helia-browser-verified-fetch/src/Output.tsx [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+          <path d="M100 50.5908C100 78.2051 77.6142 100.591 50 100.591C22.3858 100.591 0 78.2051 0 50.5908C0 22.9766 22.3858 0.59082 50 0.59082C77.6142 0.59082 100 22.9766 100 50.5908ZM9.08144 50.5908C9.08144 73.1895 27.4013 91.5094 50 91.5094C72.5987 91.5094 90.9186 73.1895 90.9186 50.5908C90.9186 27.9921 72.5987 9.67226 50 9.67226C27.4013 9.67226 9.08144 27.9921 9.08144 50.5908Z" fill="currentColor" />
+          <path d="M93.9676 39.0409C96.393 38.4038 97.8624 35.9116 97.0079 33.5539C95.2932 28.8227 92.871 24.3692 89.8167 20.348C85.8452 15.1192 80.8826 10.7238 75.2124 7.41289C69.5422 4.10194 63.2754 1.94025 56.7698 1.05124C51.7666 0.367541 46.6976 0.446843 41.7345 1.27873C39.2613 1.69328 37.813 4.19778 38.4501 6.62326C39.0873 9.04874 41.5694 10.4717 44.0505 10.1071C47.8511 9.54855 51.7191 9.52689 55.5402 10.0491C60.8642 10.7766 65.9928 12.5457 70.6331 15.2552C75.2735 17.9648 79.3347 21.5619 82.5849 25.841C84.9175 28.9121 86.7997 32.2913 88.1811 35.8758C89.083 38.2158 91.5421 39.6781 93.9676 39.0409Z" fill="currentFill" />
+        </svg>
```
Comment (SgtPooki) [inline examples/helia-browser-verified-fetch/src/Output.tsx:39]: I would prefer this as it's own svg pulled in as a component.. but not a big deal

File: examples/helia-browser-verified-fetch/test/blockstore.js [hunk 7]
```diff
@@ -0,0 +1,52 @@
+import { CodeError } from '@libp2p/interface'
+import { base32 } from 'multiformats/bases/base32'
+import { CID } from 'multiformats/cid'
+import * as raw from 'multiformats/codecs/raw'
+import * as Digest from 'multiformats/hashes/digest'
+import fixtures from './fixtures.js'
+
+export class MemoryBlockstore {
```
Comment (SgtPooki) [inline examples/helia-browser-verified-fetch/test/blockstore.js:8]: only non-nit:

why do we need this custom memory store again? could we add a link to the comments/discussion around this and some explainers in the jsdoc?

can we `export class MemoryBlockstore extends MemoryBlockstoreOriginal` so we can remove a lot of this code that detracts from the example?

**Review by SgtPooki — APPROVED** [review]
I'm good with merging this. some suggestions for improvement though

**Comment by SgtPooki** [comment]
FYI that the test succeeds locally, and when ran in isolation, but failed in monorepo test: https://github.com/ipfs-examples/helia-examples/actions/runs/8455093871/job/23162709742?pr=285#step:7:464

I kicked off another retry (but saw that it had already been retried). not sure what's going on.

File: examples/helia-browser-verified-fetch/test/blockstore.js (hunk 7 above)
Comment (2color) [reply examples/helia-browser-verified-fetch/test/blockstore.js:8]: Great idea! In fact, there's a lot of code duplication there.

File: examples/helia-browser-verified-fetch/src/Output.tsx (hunk 6 above)
Comment (2color) [reply examples/helia-browser-verified-fetch/src/Output.tsx]: Done!

**Comment by 2color** [comment]
Thanks @SgtPooki. I've addressed all the feedback.

Thrilled to have this example to point folks to.

**Pull request merged**